go-controller/pkg/ovn/master.go

```diff
@@ -121,6 +124,81 @@ func (oc *Controller) Start(kClient kubernetes.Interface, nodeName string, wg *s
 	return nil
 }
 
+// delete obsoleted logical OVN entities that are specific for Multiple join switches OVN topology. Also cleanup
+// OVN entities for deleted nodes (similiar to syncNodes() but for obsoleted Multiple join switches OVN topology
+func (oc *Controller) upgradeToSingleSwitchOVNTopology(existingNodeList *kapi.NodeList) error {
+
+	existingNodes := make(map[string]bool)
+	for _, node := range existingNodeList.Items {
+		existingNodes[node.Name] = true
+
+		// delete the obsoleted node-join-subnets annotation
+		err := oc.kube.SetAnnotationsOnNode(&node, map[string]interface{}{"k8s.ovn.org/node-join-subnets": nil})
+		if err != nil {
+			klog.Errorf("Failed to remove node-join-subnets annotation for node %s", node.Name)
+		}
+	}
+
+	nodeSwitches, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=name,other-config", "find", "logical_switch", "other-config:subnet!=_")
+	if err != nil {
+		return fmt.Errorf("failed to get node logical switches: stderr: %q, error: %v",
+			stderr, err)
+	}
+
+	// find all nodes that OVN logical entities were created for
+	logicalNodes := make(map[string]bool)
+	for _, result := range strings.Split(nodeSwitches, "\n\n") {
+		// Split result into name and other-config
+		items := strings.Split(result, "\n")
+		if len(items) != 2 || len(items[0]) == 0 {
+			continue
+		}
+		logicalNodes[items[0]] = true
+	}
+
+	for nodeName := range logicalNodes {
+		// if the node was deleted when ovn-master was down, delete its per-node switch
+		if _, ok := existingNodes[nodeName]; !ok {
+			_ = oc.deleteNodeLogicalNetwork(nodeName)
+		}
+
+		// for all nodes include the onces that were deleted or still exist, delete its gateway entities.
+		// If the node still exists, its single join switch topology logical entities will be recreated later.
+		err = multiJoinSwitchGatewayCleanup(nodeName)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (oc *Controller) upgradeOVNTopology(existingNodes *kapi.NodeList) error {
+	// Find out the current OVN topology version, if "k8s-ovn-topo-version" external_ids is does not exist, it is pre
+	// OVN topology versioning, set version number to 0
+	ver := 0
+	stdout, stderr, err := util.RunOVNNbctl("get", "logical_router", ovnClusterRouter, "external_ids:k8s-ovn-topo-version")
+	if err != nil {
+		if strings.Contains(stderr, "no row") {
+			// no ovnClusterRouter exists, DB is empty, nothing to upgrade
+			ver = OvnCurrentTopologyVersion
+		}
```

Comment:
what about an else here? log a warning? "unable to detect k8s-ovn-topo-version, will assume upgrade is needed"

Reply:
sure. I will log an info instead of warning though, since it is expected as this is the first time we introduce k8s-ovn-topo-version.

Reply:
can you also add that logging message here please in the next patch set?

Reply:
sorry that I missed this. Added now.